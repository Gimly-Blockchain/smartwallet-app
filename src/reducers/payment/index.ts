import { AnyAction } from 'redux'
import { ITransactionDetailsPaymentRequest } from 'jolocom-lib/js/interactionTokens/interactionTokens.types'

export interface PaymentStateSummary {
  readonly transactionDetails:  ITransactionDetailsPaymentRequest
  readonly requestJWT: string
  readonly description: string
  readonly didRequester: string
  readonly callbackURL?: string
}

export interface PaymentState {
  activePaymentRequest: PaymentStateSummary
}

const initialState: PaymentState = {
  activePaymentRequest: {
    requestJWT: '',
    transactionDetails: {
      receiverAddress: '',
      amountInEther: ''
    },
    description: '',
    didRequester: ''
  }
} 

export const paymentReducer = (state = initialState, action: AnyAction ): PaymentState => {
  switch(action.type) {
    case 'SET_PAYMENT_REQUEST':
      return { activePaymentRequest: action.value }
    case 'CLEAR_PAYMENT_REQUEST':
      return initialState
    default:
      return state
  }
}
