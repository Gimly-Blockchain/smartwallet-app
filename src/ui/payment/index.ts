export { PaymentConsent } from 'src/ui/payment/containers/paymentConsent'