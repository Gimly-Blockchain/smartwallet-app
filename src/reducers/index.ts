import { combineReducers } from 'redux'
import { settingsReducer, SettingsState } from 'src/reducers/settings/'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  settings: settingsReducer,
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly settings: SettingsState
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
