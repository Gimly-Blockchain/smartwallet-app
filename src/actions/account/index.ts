import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import { CategorizedClaims, DecoratedClaims } from 'src/reducers/account'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
} from '../../lib/util'
import { cancelReceiving } from '../sso'
import { JolocomLib } from 'jolocom-lib'
import { ThunkAction } from 'src/store'
import { groupBy, map, mergeRight, omit, uniq, zipWith } from 'ramda'
import { compose } from 'redux'
import { CredentialMetadataSummary } from '../../lib/storage/storage'
import { IdentitySummary } from '../sso/types'
import { KeyTypes } from 'jolocom-lib/js/vaultedKeyProvider/types'
import { publicKeyToDID } from 'jolocom-lib/js/utils/crypto'
import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { Not } from 'typeorm'
import { HAS_EXTERNAL_CREDENTIALS } from './actionTypes'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANDLE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const checkIdentityExists: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { keyChainLib, storageLib, encryptionLib, registry } = backendMiddleware
  const encryptedEntropy = await storageLib.get.encryptedSeed()

  if (!encryptedEntropy) {
    const isRegistering = getState().registration.loading.isRegistering

    const routeName = isRegistering
      ? routeList.RegistrationProgress
      : routeList.Landing

    return dispatch(navigationActions.navigate({ routeName }))
  }

  const encryptionPass = await keyChainLib.getPassword()

  const decryptedSeed = encryptionLib.decryptWithPass({
    cipher: encryptedEntropy,
    pass: encryptionPass,
  })

  if (!decryptedSeed) {
    throw new Error('could not decrypt seed')
  }

  // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
  const userVault = JolocomLib.KeyProvider.fromSeed(
    Buffer.from(decryptedSeed, 'hex'),
    encryptionPass,
  )

  const userPubKey = userVault.getPublicKey({
    derivationPath: KeyTypes.jolocomIdentityKey,
    encryptionPass,
  })

  const didDocument = await storageLib.get.didDoc(publicKeyToDID(userPubKey))

  if (didDocument) {
    const identity = Identity.fromDidDocument({ didDocument })

    // TODO Simplify constructor
    backendMiddleware.identityWallet = new IdentityWallet({
      identity,
      vaultedKeyProvider: userVault,
      publicKeyMetadata: {
        derivationPath: KeyTypes.jolocomIdentityKey,
        keyId: identity.publicKeySection[0].id,
      },
      contractsAdapter: registry.contractsAdapter,
      contractsGateway: registry.contractsGateway,
    })
  } else {
    const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes
    const identityWallet = await registry.authenticate(userVault, {
      encryptionPass,
      derivationPath,
    })

    backendMiddleware.identityWallet = identityWallet
    await storageLib.store.didDoc(identityWallet.didDocument)
  }

  const userDid = backendMiddleware.identityWallet.identity.did
  dispatch(setDid(userDid))

  return dispatch(navigationActions.navigate({ routeName: routeList.Home }))
}

export const openClaimDetails = (
  claim: DecoratedClaims,
): ThunkAction => dispatch => {
  dispatch(setSelected(claim))
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { identityWallet, storageLib, keyChainLib } = backendMiddleware

  const did = getState().account.did.did
  const claimsItem = getState().account.claims.selected
  const password = await keyChainLib.getPassword()

  const verifiableCredential = await identityWallet.create.signedCredential(
    {
      metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
      // the library acts directly on the object passed in, so a copy should be made first
      claim: { ...claimsItem.claimData },
      subject: did,
    },
    password,
  )

  if (claimsItem.id) {
    await storageLib.delete.verifiableCredential(claimsItem.id)
  }

  await storageLib.store.verifiableCredential(verifiableCredential)

  await dispatch(setClaimsForDid)

  return dispatch(navigationActions.navigatorResetHome())
}

// TODO Currently only rendering  / adding one
export const saveExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib } = backendMiddleware
  const externalCredentials = getState().account.claims.pendingExternal

  if (!externalCredentials.offer.length) {
    return dispatch(cancelReceiving)
  }

  const cred: SignedCredential = externalCredentials.offer[0].credential

  await storageLib.delete.verifiableCredential(cred.id)
  await storageLib.store.verifiableCredential(cred)

  return dispatch(cancelReceiving)
}

export const toggleLoading = (value: boolean) => ({
  type: 'SET_LOADING',
  value,
})

export const hasExternalCredentials: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib, identityWallet } = backendMiddleware
  const externalCredentials = await storageLib.get.verifiableCredential({
    issuer: Not(identityWallet.did),
  })

  return dispatch({
    type: HAS_EXTERNAL_CREDENTIALS,
    value: externalCredentials.length !== 0,
  })
}
export const setClaimsForDid: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  const { storageLib } = backendMiddleware

  const verifiableCredentials: SignedCredential[] = await storageLib.get.verifiableCredential()

  const metadata = await Promise.all(
    verifiableCredentials.map(el => storageLib.get.credentialMetadata(el)),
  )

  const issuers = uniq(verifiableCredentials.map(cred => cred.issuer))

  const issuerMetadata = await Promise.all(
    issuers.map(storageLib.get.publicProfile),
  )

  const claims = prepareClaimsForState(
    verifiableCredentials,
    metadata,
    issuerMetadata,
  ) as CategorizedClaims

  return dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })
}

export const prepareClaimsForState = (
  credentials: SignedCredential[],
  credentialMetadata: Array<CredentialMetadataSummary | {}>,
  issuerMetadata: Array<IdentitySummary | { did: string }>,
) =>
  compose(
    groupBy(getCredentialUiCategory),
    zipWith(mergeRight, credentialMetadata),
    map(addIssuerInfo(issuerMetadata)),
    map(convertToDecoratedClaim),
  )(credentials)

export const addIssuerInfo = (
  issuerProfiles: Array<{ did: string } | IdentitySummary> | [],
) => (claim: DecoratedClaims) => {
  if (!issuerProfiles || !issuerProfiles.length) {
    return claim
  }

  const issuer = issuerProfiles.find(el => el.did === claim.issuer.did)

  return issuer
    ? {
        ...claim,
        issuer,
      }
    : claim
}

/** @TODO Util, make subject mandatory (in lib) */
export const convertToDecoratedClaim = ({
  claim,
  type,
  issuer,
  id,
  expires,
}: SignedCredential): DecoratedClaims => ({
  credentialType: getUiCredentialTypeByType(type),
  issuer: {
    did: issuer,
  },
  claimData: omit(['id'], claim),
  id,
  subject: claim.id || 'Not found',
  expires: expires || undefined,
})
