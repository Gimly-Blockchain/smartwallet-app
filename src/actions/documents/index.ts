import { Action, AnyAction } from 'redux'
import { Dispatch } from 'react-redux'
import { navigationActions } from '..'
import { routeList } from 'src/routeList'
import I18n from 'src/locales/i18n'
import { DecoratedClaims } from 'src/reducers/account'
import { ClaimInterface } from 'cred-types-jolocom-core'

export const SET_EXPIRED_DOC = 'SET_SELECTED_EXPIRED_DOCUMENT'
export const CLEAR_EXPIRED_DOC = 'CLEAR_SELECTED_EXPIRED_DOCUMENT'

export const setSelectedExpiredDocument = (
  document: DecoratedClaims,
): AnyAction => ({
  type: SET_EXPIRED_DOC,
  value: document,
})

export const clearSelectedExpiredDocument = (): Action => ({
  type: CLEAR_EXPIRED_DOC,
})

export const openExpiredDetails = (document: DecoratedClaims) => async (
  dispatch: Dispatch<AnyAction>,
) => {
  dispatch(setSelectedExpiredDocument(document))
  dispatch(
    navigationActions.navigate({
      routeName: routeList.ExpiredDetails,
      params: {
        headerTitle: `${I18n.t('[expired]')} ${
          (document.claimData as ClaimInterface).type
        }`,
      },
    }),
  )
}
