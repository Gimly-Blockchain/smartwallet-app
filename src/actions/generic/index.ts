import { Linking } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { routeList } from 'src/routeList'
import { navigationActions, accountActions } from 'src/actions'
import { ThunkAction } from 'src/store'
import settingKeys from 'src/ui/settings/settingKeys'
import { withLoading, withErrorScreen } from '../modifiers'
import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
} from 'src/reducers/generic'
import { AnyAction } from 'redux'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300
import { AppError, ErrorCode } from '@jolocom/sdk/js/src/lib/errors'
import { AsyncStorage } from 'react-native'
import crypto from 'crypto'
import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()

    // locale setup
    /** locale setup
     * @dev Until German and Dutch terms are polished, only English is used.
     * previous code:
     */
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale

    storedSettings.locale = I18n.locale

    await dispatch(loadSettings(storedSettings))

    const ret = await dispatch(accountActions.checkIdentityExists)

    // FIXME what happens if no identity and this is a deeplink?
    // navigationActions.handleDeepLink throws NoWallet
    // need to improve this UX

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    await Linking.getInitialURL().then(url => {
      if (url) handleDeepLink(url)
    })

    return ret
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  } finally {
    SplashScreen.hide()
  }
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()
  return dispatch(
    navigationActions.navigate({
      routeName: shouldShowTerms ? routeList.TermsOfServiceConsent : route,
      params: { nextRoute: route, onSubmit },
    }),
  )
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: 'LOAD_SETTINGS',
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
