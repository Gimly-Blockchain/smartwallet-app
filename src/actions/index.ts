import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as interactionHandlerActions from 'src/actions/interactionHandler'
import * as paymentActions from 'src/actions/payment/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  interactionHandlerActions,
  paymentActions
}
