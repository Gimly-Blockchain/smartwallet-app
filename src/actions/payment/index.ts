import { Dispatch, AnyAction } from 'redux'
import { JolocomLib } from 'jolocom-lib'
import { BackendMiddleware } from 'src/backendMiddleware'
import { navigationActions, accountActions } from 'src/actions'
import { showErrorScreen } from 'src/actions/generic'
import { JSONWebToken } from 'jolocom-lib/js/interactionTokens/JSONWebToken'
import { PaymentRequest} from 'jolocom-lib/js/interactionTokens/paymentRequest'
import {Â jolocomEthTransactionConnector } from 'jolocom-lib/js/ethereum/transactionConnector'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { publicKeyToAddress } from  'jolocom-lib/js/utils/helper'
import { PaymentStateSummary } from 'src/reducers/payment'
import { routeList } from 'src/routeList'
import { Linking } from 'react-native'

export const setPaymentRequest = (request: PaymentStateSummary) => {
  return {
    type: 'SET_PAYMENT_REQUEST',
    value: request
  }
}

export const clearPaymentRequest = () => {
  return {
    type: 'CLEAR_PAYMENT_REQUEST'
  }
}

export const cancelPaymentRequest = () => {
  return (dispatch: Dispatch<AnyAction>, getState: Function) => {
    const { callbackURL } = getState().payment.activePaymentRequest
    dispatch(clearPaymentRequest())

    /**
     * This needs to be cleaned up after demo
     * TODO: consolidate routing back to callbackURL for deep linking for payment & sso
     */
    if (callbackURL.includes('http')) {
      dispatch(navigationActions.navigatorReset({ routeName: routeList.Home }))
    } else {
      const url = callbackURL + Buffer.from(JSON.stringify({token: false})).toString('base64')
      Linking.openURL(url)
    }
  }
}

export const consumePaymentRequest = (paymentRequest: JSONWebToken<PaymentRequest>) => {
  return async (dispatch: Dispatch<AnyAction>, getState: Function, backendMiddleware: BackendMiddleware) => {
    const { identityWallet } = backendMiddleware

    try {
      await identityWallet.validateJWT(paymentRequest)

      const summary = {
        didRequester: paymentRequest.issuer, // TODO: replace with public profile data later
        description: paymentRequest.interactionToken.description,
        transactionDetails: paymentRequest.interactionToken.transactionDetails,
        requestJWT: paymentRequest.encode()
      }

      dispatch(setPaymentRequest(summary))
      dispatch(accountActions.toggleLoading(false))
      dispatch(navigationActions.navigate({ routeName: routeList.PaymentConsent }))
    } catch (err) {
      dispatch(accountActions.toggleLoading(false))
      dispatch(showErrorScreen(new Error('Consuming payment request failed.: ')))
    }
  }
}

export const sendPaymentResponse = () => {
  return async (dispatch: Dispatch<AnyAction>, getState: Function, backendMiddleware: BackendMiddleware) => {
    const { keyChainLib, identityWallet, storageLib, encryptionLib } = backendMiddleware

    const { activePaymentRequest } = getState().payment
    const paymentRequest = JolocomLib.parse.interactionToken.fromJWT(activePaymentRequest.requestJWT)

    try {
      const password = await keyChainLib.getPassword()

      const ethAddress = publicKeyToAddress(identityWallet.getPublicKey({
        encryptionPass: password,
        derivationPath: JolocomLib.KeyTypes.ethereumKey
      }))

      const tx = await jolocomEthTransactionConnector.createTransaction({
        ...paymentRequest.interactionToken.transactionDetails,
        senderAddress: ethAddress
      })

      const decryptedSeed = encryptionLib.decryptWithPass({
        cipher: await storageLib.get.encryptedSeed(),
        pass: password
      })

      const userVault = new SoftwareKeyProvider(Buffer.from(decryptedSeed, 'hex'), password)

      tx.sign(userVault.getPrivateKey({
        encryptionPass: password,
        derivationPath: JolocomLib.KeyTypes.ethereumKey
      }))
  
      const txReceipt = await jolocomEthTransactionConnector.sendSignedTransaction(tx.serialize())

      const paymentResponseJWT = await identityWallet.create.interactionTokens.response.payment(
        { txHash: txReceipt.hash },
        password,
        paymentRequest
      )

      const { callbackURL } = paymentRequest.interactionToken
      // TODO: introduce generic handling linking vs https
      if (callbackURL.includes('http')) {
        await fetch(callbackURL, {
          method: 'POST',
          body: JSON.stringify({ token: paymentResponseJWT.encode() }),
          headers: { 'Content-Type': 'application/json' }
        })
      } else {
        const url = callbackURL + paymentResponseJWT.encode()
        Linking.openURL(url)
      }

      dispatch(clearPaymentRequest())
      dispatch(accountActions.toggleLoading(false))
      dispatch(navigationActions.navigate({ routeName: routeList.Home }))
    } catch (err) {
      dispatch(accountActions.toggleLoading(false))
      dispatch(showErrorScreen(new Error('Creating and sending payment response failed.')))
    }
  }
}
