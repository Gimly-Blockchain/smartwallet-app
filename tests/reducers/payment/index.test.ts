import { paymentReducer } from 'src/reducers/payment'
import { paymentActions as actions } from 'src/actions/'

describe('payment reducer', ()=> {
  it('should initialize correctly', () => {
    expect(paymentReducer(undefined, { type: '@INIT' })).toMatchSnapshot()
  })

  it('should handle the SET_PAYMENT_REQUEST action ', () => {
    expect(paymentReducer(undefined, actions.setPaymentRequest({
      transactionDetails: {
        receiverAddress: 'ethereumAddress',
        amountInEther: '0.1'
      },
      requestJWT: 'zfghik.zfghj.rtdaftgz',
      description: 'This is a payment example',
      didRequester: 'did:jolo:test'
    })))
      .toMatchSnapshot()
  })

  it('should handle the CLEAR_PAYMENT_REQUEST action ', () => {
    expect(paymentReducer(undefined, actions.clearPaymentRequest()))
      .toMatchSnapshot()
  })
})
