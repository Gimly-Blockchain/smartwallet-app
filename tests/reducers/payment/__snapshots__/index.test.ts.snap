// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payment reducer should handle the CLEAR_PAYMENT_REQUEST action  1`] = `
Object {
  "activePaymentRequest": Object {
    "description": "",
    "didRequester": "",
    "requestJWT": "",
    "transactionDetails": Object {
      "amountInEther": "",
      "receiverAddress": "",
    },
  },
}
`;

exports[`payment reducer should handle the SET_PAYMENT_REQUEST action  1`] = `
Object {
  "activePaymentRequest": Object {
    "description": "This is a payment example",
    "didRequester": "did:jolo:test",
    "requestJWT": "zfghik.zfghj.rtdaftgz",
    "transactionDetails": Object {
      "amountInEther": "0.1",
      "receiverAddress": "ethereumAddress",
    },
  },
}
`;

exports[`payment reducer should initialize correctly 1`] = `
Object {
  "activePaymentRequest": Object {
    "description": "",
    "didRequester": "",
    "requestJWT": "",
    "transactionDetails": Object {
      "amountInEther": "",
      "receiverAddress": "",
    },
  },
}
`;
